File: romberg.py
import numpy as np

def f(x):
   'Función que se desea integrar'
   f=1/x
   return f

def romberg(a, b, k, l):
   '''
   Esta función realiza integración numérica por medio del método de Romberg(extrapolación de Richardson)
   Inputs:
      a: limite inferior de integración
      b: limite superior de integración
      k: 1,2,3, n
      l: 1,2,..., k
   Output:
      r: valor de la integral 
   '''
   if l>k:
      print('Error, j es mayor a i')
      return 
   if l==0 or k==0:
      print('Error, i y j tienen que ser mayores a cero')
      return

   r=np.zeros((k,l)) #array donde se guardan los valores de rij
   sum=0
   for i in range(1,k+1):
      for j in range(1,l+1):
         if ( i==1 and j==1 ):   #Calcula R11
            h = (b-a)/i
            r[0,0] = h/2 *( f(a) + f(b) )

         elif ( i>=2 and j==1 ):   #Calcula Ri1
            for m in range(1,2**(i-2)+1): 
               sum=sum + f(a + (2*m-1)*h/2 ) #sumatoria 
            r[i-1,0] = 0.5 * ( r[i-2,0] + h*sum )
            h = (b-a)/i
            sum=0

         else: #Calcula Rij, i>=2 j>2
            r[i-1,j-1] = ( 4**(j-1)*r[i-1,j-2]-r[i-2,j-2] )/( 4**(j-1)-1 )

   return r[k-1,l-1]

print(romberg(1,3,3,3))

File: romberg.f90
program main
   implicit none
   real::   romberg
   write(*,*) romberg(1.0,3.0,3,3)
end program main

real function f(x) !Función qu se desea integrar
   implicit none
   real:: x

   f=1/x
end function f

real function romberg(a, b, i, j)
!Esta función realiza integración numérica por medio del método de Romberg(extrapolación de Richardson)
!Inputs:
!   a: limite inferior de integración
!   b: limite superior de integración
!   i: 1,2,3, ... , n
!   j: 1,2,3, ... , i
!Output:
!   r: valor de la integral 
   implicit none
   real  :: f, a, b, sum, h
   integer  :: i, j, k, l, m
   real  :: r(i,j)


   if (j>i) then
      write(*,*) 'Error, j es mayor a i'
      return 
   end if

   if (i<=0 .or. j<=0) then
      write(*,*) 'Error, i y j tienen que ser mayores a cero'
      return
   end if

   sum=0 !se almacena el valor de la sumatoria
   do l=1,j !ciclo para movernos entre columnas
      do k=1,i !ciclo para movernos entre filas

         if ( k<l ) cycle !solamente calculamos los elementos de la matriz triangular

         if ( k==1 .and. l==1 ) then   !calcula r11
            h = (b-a)/k
            r(1,1) = h/2 *( f(a) + f(b) )

         else if ( k>=2 .and. l==1 ) then !calcula ri1
            do m=1, 2**(k-2)
               sum = sum + f(a + (2*m-1)*h/2 )
            end do
            r(k,1) =  0.5 * ( r(k-1,1) + h*sum )
            h = (b-a)/k
            sum=0

         else  !calcula rij, i>=2 j>2
            r(k,l) = ( 4**(l-1)*r(k,l-1)-r(k-1,l-1) )/( 4**(l-1)-1 )
         end if

      end do
   end do
   romberg=r(i,j)
end function romberg